SRE-49 - fix windows wallet tests in hydra

jbgi opened the pull request:
### Issue

[SRE-49](https://jira.iohk.io/browse/SRE-49)

### Summary

Fixes windows tests under wine by:
1. Updating the [wine version to 3.2.1](https://github.com/input-output-hk/haskell.nix/pull/472) and the [windows testWrapper](https://github.com/input-output-hk/haskell.nix/pull/471) so that we can run jormungandr under wine (allows integration tests to run under wine).
2. Fix utf-8 encoding issue in cli specs.
3. Fix errors found when deleting a wallet while its database is currently open.
4. In `DBFactory.removeDatabase`, only delete the file once there are no `withDatabase` actions in progress.

### Comments

- [x] [cardano-wallet-cli.unit](https://hydra.iohk.io/job/Cardano/cardano-wallet-pr-1373/x86_64-w64-mingw32.checks.cardano-wallet-cli.unit.x86_64-linux)
- [x] [cardano-wallet-core.unit](https://hydra.iohk.io/job/Cardano/cardano-wallet-pr-1373/x86_64-w64-mingw32.checks.cardano-wallet-core.unit.x86_64-linux)
- [ ] [cardano-wallet-jormungandr.integration](https://hydra.iohk.io/job/Cardano/cardano-wallet-pr-1373/x86_64-w64-mingw32.checks.cardano-wallet-jormungandr.integration.x86_64-linux)
- [ ] [cardano-wallet-launcher.unit](https://hydra.iohk.io/job/Cardano/cardano-wallet-pr-1373/x86_64-w64-mingw32.checks.cardano-wallet-launcher.unit.x86_64-linux) ⇒ #1425

jbgi committed:
Use haskell.nix branch with wine fixes

jbgi committed:
Try delete multiple times on windows

 as workaround for https://github.com/haskell/directory/issues/96

jbgi committed:
fix "removeDatabase works and remove files" test to work on windows.

jbgi committed:
Fix test failure (encoding error) on windows by forcing utf-8

KtorZ commented on lib/cli/test/unit/Cardano/CLISpec.hs:
```diff
@@ -127,17 +128,21 @@ spec = do
-    let expectStdErr args expectation = it (unwords args) $
+    let expectStdErr args expectation = it (unwords args) $ do
+            hSetEncoding IO.stdout utf8
+            hSetEncoding IO.stderr utf8
```
@jbgi you may want to use: https://input-output-hk.github.io/cardano-wallet/haddock/edge/cardano-wallet-launcher-2020.3.3/Cardano-Startup.html#v:withUtf8Encoding (for a cross-platform solution to this). We've ran into this a couple of times already :(

jbgi replied:
ah nice!

jbgi replied:
`withUtf8Encoding` did not have any effect on the test failing (presumably due to `withCP65001` logic), had to use `setUtf8EncodingHandles` directly.

KtorZ commented on lib/core/src/Cardano/Wallet/DB/Sqlite.hs:
```diff
@@ -281,6 +298,36 @@ newDBFactory tr defaultFieldValues mDatabaseDir = do
+    retryOnPermissionErrors 5 . removePathForcibly
+#else
+    removePathForcibly
+#endif
```
We'd need to check if there's a way to check / make sure that the SQLite has been properly closed before removing the file. In practice, we always:

a. Kill the thread that has the connection open (which, because of bracketting-style resource allocation should free-up the connection at some point)

b. Only remove the database file when we know that the connection has ended. 

On Linux, it seems that either the connection is released fast enough or, that it doesn't matter if the connection is still open when the file is deleted. On Windows, that's a different story indeed. 

One possible "easy way" to work around this I see would be to have an MVar shared between the `withDatabase` and `removeDatabase` such that, `removeDatabase` would only attempt to remove the file once the connection has been terminated on `withDatabase`; so waiting on an MVar before `mapM_ removePathForcibly ...` with a protective time-out. 

cc @rvl

rvl replied:
@jbgi This is interesting. What references do you have for these claims?

1. "On Windows ... a FileDelete command just marks a file for deletion. The file is really only removed when the last handle to the file is closed."
2. "there are a lot of system services that can have a file temporarily opened using a shared read-only lock, such as the built in AV and search indexer."

For 1, I thought that was more like the behaviour on Linux (although on Linux, the directory entry is immediately removed, but the inode stays). Do you mean that the filename lingers in the directory listing untll it has actually been deleted?

For 2, do these services _really_ cause files to not be deleted promptly? I thought that would break lots of things that expect FileDelete to do the obvious.

jbgi replied:
main reference: https://github.com/haskell/directory/issues/96
I originally implemented this "fix" in order to try getting a test pass on wine, but in the end, it was not necessary for the test to pass, or at least not enough, though I still think this hack is useful from what I have read.

jbgi committed:
Delete empty file

rvl commented on nix/sources.json:
```diff
@@ -1,26 +1,26 @@
 {
     "haskell.nix": {
-        "branch": "master",
+        "branch": "jbgi/cardano-wallet-fix-windows",
```
This needs to be master branch, so your PRs on Haskell.nix need to be merged first.

jbgi replied:
agreed, though cardano-wallet master is currently not using haskell.nix master but https://github.com/input-output-hk/haskell.nix/tree/hkm/cardano-wallet-fix

rvl replied:
cardano-wallet master is on haskell.nix master I'm pretty sure.

jbgi replied:
It is not, see https://input-output-rnd.slack.com/archives/CG386Q1NE/p1582245051365500?thread_ts=1582215136.357100&cid=CG386Q1NE for context. (maybe the issue has been solved on master since then, cc/ @hamishmack)

rvl commented on lib/core/src/Cardano/Wallet/DB/Sqlite.hs:
```diff
@@ -281,6 +296,36 @@ newDBFactory tr defaultFieldValues mDatabaseDir = do
+                    let retries' = traceShow (ex, maxRetries) (maxRetries - 1)
+                    threadDelay ((maxRetries - retries') * 200)
+                    retryOnPermissionErrors retries' io
+                | otherwise -> traceShow ("OTHER ERROR" :: String) (throw ex)
```
If we end up using the retry approach: You can log with the iohk-monitoring tracer here, rather than Debug.Trace.traceShow. Also some debug logging on retries would help.

jbgi replied:
Yes, though is this falling outside my haskell proficiency atm :(.
There is debug logging on retries at line 323:
```haskell
 let retries' = traceShow (ex, maxRetries) (maxRetries - 1)
```

rvl replied:
OK - I will convert the logging to use Control.Tracer.

rvl reviewed (commented):
It might be better to implement some code using MVars which deletes the sqlite db file after the last connection has closed, as @KtorZ suggests.

rvl committed:
sqlite: Add logging and test case for removeDatabase retrying

rvl committed:
Appease stylish-haskell

rvl committed:
Update haskell.nix to latest master rev

rvl committed:
SQlite.DBFactory.removeDatabase: Don't delete open databases

Wait for all connections to a file to be closed before deleting a
database.

This helps avoid "file in use" errors on Windows.

rvl committed:
tests: Add unit tests for database delete functions

rvl reviewed (approved):
Remaining two test failures can be fixed in a subsequent PR.

KtorZ commented on lib/core/src/Cardano/DB/Sqlite/Delete.hs:
```diff
@@ -0,0 +1,237 @@
+data RefCount ix = RefCount
+    { _refCount :: MVar (Map ix Int) -- ^ number of references to each index
+    , _takeLock :: MVar () -- ^ lock on incrementing references
+    }
```
I don't think we do have more than one `withDatabase` call per wallet, so there should be always at most a single connection per database :thinking:. 
If we can do more, we can also do less so I guess it's fine.

rvl replied:
I thought that might be the case, but am hoping it will be more resilient if it supports more than one `withDatabase` call per wallet.

KtorZ reviewed (approved):
LGTM. Haven't tried locally but looks solid and quite convinced unit tests.

Anviking commented on lib/core/src/Cardano/Wallet/DB/Sqlite.hs:
```diff
@@ -258,28 +260,33 @@ newDBFactory tr defaultFieldValues mDatabaseDir = do
+                -- try to wait for all 'withDatabase' calls to finish before
+                -- deleting database file.
+                let trWait = contramap (MsgWaitingForDatabase widp) tr
+                waitForFree trWait refs wid $ \inUse -> do
```
Why do we need the reference counting now, but didn't previously?

Only using `deleteSqliteDatabase`, without `waitForFee`, wouldn't achieve the same thing? (which is, I guess, to make sure the file is really gone when `removeDatabase` returns)

Some more wandering questions from trying to understand:

Is `withDatabase` is called once by the restoration worker? Then I guess `waitForFree` makes sure that the wallet is not in the registry. _And_ it looks like `destroyDBLayer` makes sure to close the sqlite connection before it is removed from the registry.

So we could never cancel a sqlite query, but theoretically an API request could acquire a `SqliteContext` when it is open, wait, and when it tries to use it it has been closed?

KtorZ replied:
> Why do we need the reference counting now, but didn't previously?

We actually always needed it. It has never worked properly on Windows :)

jbgi commented on nix/sources.json:
```diff
@@ -5,22 +5,22 @@
     },
     "iohk-nix": {
-        "branch": "master",
+        "branch": "nixpkgs-common-channels",
```
need https://github.com/input-output-hk/iohk-nix/pull/342 to be merged (or rollback this change).

rvl replied:
Thanks

jbgi committed:
Use iohk-nix master branch.

rvl committed:
Fix timing mistake in waitForFree test cases

rvl committed:
Buildkite: increase time limit for tests from 30 to 45 minutes
